### PR TITLE
Display the git graph

### PR DESCRIPTION
**Pull request by jesseduffield**
How it works:
* each 'pipe' in the graph is coloured based on the author of the descendant commit. When selecting a commit, the pipes to its ancestors will be highlighted in white to remove any ambiguity.
* to get the graph looking good, we're sorting the git log in topological order as opposed to committer date. I'm pretty sure this won't have any effect on interactive rebases (those drop merges by default anyway)

When you're looking at a section of the graph that's dominated by a single author, it does make it harder to grok what's going on, but when there's a mix of authors, having the pipes coloured by author makes it much better than using random pipe colours. Not sure whether to make that setting configurable.

https://user-images.githubusercontent.com/8456633/139527896-316b9980-f604-4170-94bd-c7883ce0e902.mov

image

### CODE REVIEW

**Comment by mjarkk** [comment]
Currently when i checkout this branch i get this error:
image

**Comment by mjarkk** [comment]
> Adding a config option to turn the graph off when viewing the commits view in a minified window

So as i always have my terminal windowed this is how some of my repos look, and this will be the default behaviour when i read the above.  
I'm not sure if that's good default behaviour. 

I'm just thinking out loud here but maybe on a small screen the horizontal lines that are not of the visible commits can be hidden?  

image

Also love this feature!!

**Comment by jesseduffield** [comment]
@mjarkk whoops that logging issue would be because I've got some custom logging happening for the sake of debugging. I've removed that now

**Comment by jesseduffield** [comment]
@mjarkk good points. I'll set the graph to be hidden by default in the small view. I'm hesitant to hide pipes that begin and end outside the bounds of the view, just because you may be trying to follow those pipes from beginning to end and that would be impossible if they disappear halfway through.

I agree about spacing concerns when we are showing the graph though. I think I'll shorten the width of the date so that it doesn't include the time. I think that's okay given that the time is still shown in the main view. It sounds like we also need a way to toggle the graph when the panel is maximised, and potentially also a way to scroll horizontally in the panel. If I hadn't made left and right the original keybindings for switching panels I definitely would have used them here. Now that you can also use tab/shift-tab to cycle panels, using arrow keys for that makes more sense, but I still don't want to mess with that given how many people are familiar with it now. Maybe cmd+left and cmd+right? I'm pretty sure I can make that work.

**Comment by mjarkk** [comment]
I'm not sure if this would work but what about showing the commit as 2 lines?  
like:

```
{{ pipes }} {{ comit id }} {{ date }} {{ author }}
{{ pipes }} {{ commit message }}
```

**Comment by jesseduffield** [comment]
@mjarkk hmm I find that a little hard to read :/

**Comment by jesseduffield** [comment]
Some new changes:
* you can now scroll left-right in a list panel with H/L. When the main view is focused these keybindings also apply, but when the main view is not focused it should be wrapping so there's no need for horizontal scrolling there.
* by default the graph does not appear when the commits window is minimised. This is controlled in the config via `git.log.showGraph` (accepted values are 'always', 'never', and 'when-maximised')
* for the sake of performance I've added some throttling to the code that renders content to the main view, so that if you're flicking through things it's not slowing things down by trying to render the content of the git show command. It will probably take some tweaking to get this right: I'm currently using the start time of the command as a crude proxy for the system being under load.
* you can now view the entire commit graph (i.e. pass --all to the `git log` command via the commit log menu (opened with <c-l>). In this menu you can also determine the order of commits and whether to show the graph at all.
* By default, we now sort commits by topographical order. This is what tig uses by default and makes for a much neater looking commit graph. You can change the order via `git.log.order` or from within lazygit in the abovementioned menu (but that won't persist it between sessions).

Because of how we render our views, there is still an annoyingly long load if you're going past 200 commits. I spent some time experimenting with an approach that just rendered what you needed to the view instead of rendering the whole e.g. 30k commit history at once, but couldn't find a nice way of handling searches in the view given that currently search applies to the view's content itself. I'm not sure how often somebody goes back beyond 200 commits but it's something we should make more performant in the future.

**Comment by jesseduffield** [comment]
@mjarkk @Ryooooooga @Liberatys @samburville (or anyone else watching) if anybody's interested in this feature lemme know your thoughts. I'm happy to merge at this point

**Comment by mjarkk** [comment]
Just tested this feature on a linux machine and almost all fonts i try have a size mismatch of the `⎔` character and the bold version **`⎔`**:
image
It's a bit anoying to see characters change in size when scrolling

**Comment by jesseduffield** [comment]
@mjarkk interesting, looks like we'll need to pick new symbols then :/

**Comment by mjarkk** [comment]
The `⬡` character doesn't seem to have the same problems maybe this one is usable

**Comment by jesseduffield** [comment]
funnily I get no problems with the original character but `⬡` actually does give me problems

**Comment by jesseduffield** [comment]
I've switched to using `◯`. Could you give that a go @mjarkk ?

**Comment by mjarkk** [comment]
The character size looks good now.  
Tough now the ⏣ looks a bit wired compared to the circles  
What about using ⦾, ⦿ or ⊙ in staid of ⏣?

**Review by Liberatys — APPROVED** [review]
Really like the feature :D 👍🏻 Thank you for all your efforts 🙏🏻

File: pkg/gui/list_context.go [hunk 1]
```diff
@@ -139,8 +159,44 @@ func (self *ListContext) handleNextLine() error {
...
+	// get the view, move the origin
+	view, err := self.Gui.g.View(self.ViewName)
+	if err != nil {
+		return nil
+	}
+
+	self.Gui.scrollRight(view)
+
+	return self.HandleFocus()
+}
```
[the remark this answers is not in the document]
Comment (jesseduffield) [reply pkg/gui/list_context.go]: fixed!

**Comment by jesseduffield** [comment]
@mjarkk good suggestions, I've given that a go but I actually prefer the merge commits being larger because it makes them easier to spot.

I'm happy to merge this now and see how people respond to it. Any objections?

**Comment by lei4519** [comment]
> * you can now view the entire commit graph (i.e. pass --all to the `git log` command via the commit log menu (opened with ). In this menu you can also determine the order of commits and whether to show the graph at all.

@jesseduffield I prefer to view whole git graph, but now I have to switch every time. Is there any way to make `—all` the default behavior? looking forward to your reply.

**Comment by jesseduffield** [comment]
@lei4519 yep, in your config go:
```yaml
git:
  log:
    showGraph: 'always'
```

**Comment by lei4519** [comment]
@jesseduffield  I'm sorry I didn't make myself clear.
When I go into the commits panel, it looks like this: 
image

I hit 'ctrl-l' and select 'toggle show whole git graph', it looks like this: 
image

I want the commits panel to look like the last image by default.

**Comment by jesseduffield** [comment]
Ah I see. That's not currently supported, could you raise an issue for that?

**Comment by lei4519** [comment]
sure, It's here https://github.com/jesseduffield/lazygit/issues/1951

**Pull request merged**